### PR TITLE
Introduction of build flags for tests

### PR DESCRIPTION
**Pull request by ruflin**
Every test file must now have a build flag, either unit or integration. Further test flags like benachmark can be added later if needed.

Before integration tests included unit tests. This is not the case anymore as now the test can be differentiated from each other. This also means testing.Short() is not needed anymore.

Closes https://github.com/elastic/beats/issues/950

### CODE REVIEW

File: libbeat/scripts/Makefile [hunk 1]
```diff
@@ -116,17 +116,17 @@ prepare-tests:
 # Race is not enabled for unit tests because tests run much slower.
 .PHONY: unit-tests
 unit-tests: prepare-tests
-	$(GOPATH)/bin/gotestcover -coverprofile=${COVERAGE_DIR}/unit.cov -short ${GOPACKAGES}
+	$(GOPATH)/bin/gotestcover -tags=unit $(RACE) -coverprofile=${COVERAGE_DIR}/unit.cov  ${GOPACKAGES}
```
Comment (ruflin) [inline libbeat/scripts/Makefile]: @urso I added the race flag to our unit-tests as I think it was actually missing. Less fun fact is that we seem to have some race conditions in libbeat: http://build-eu-00.elastic.co/job/Beats-PR/746/console

Comment (ruflin) [inline libbeat/scripts/Makefile]: @urso I will remove the race flag again so we can get this PR in without fixing this. But I opened a separate issue to keep track on it: https://github.com/elastic/beats/issues/1074

File: libbeat/common/datetime_test.go [hunk 2]
```diff
@@ -41,7 +43,7 @@ func TestParseTime(t *testing.T) {
 		{
 			Input:  "2015-04-31T11:19:05.112Z",
 			Output: time.Date(2015, time.April, 31, 11, 19, 05, 112*1e6, time.UTC),
-		},
+		},*/
```
Comment (tsg) [inline libbeat/common/datetime_test.go]: any reason for disabling these?

**Comment by tsg** [comment]
LGTM.

File: libbeat/common/datetime_test.go (hunk 2 above)
Comment (ruflin) [inline libbeat/common/datetime_test.go]: problem with golang 1.6. They were now removed anyways in https://github.com/elastic/beats/pull/1042, hence the required rebase.

Comment (urso) [inline libbeat/common/datetime_test.go]: these have been 'bugs' fixed in  go 1.6 .

**Comment by ruflin** [comment]
@tsg Rebased on top of master.

**Comment by tsg** [comment]
Ok, waiting for green.

File: libbeat/outputs/elasticsearch/api_test.go [hunk 3]
```diff
@@ -1,3 +1,5 @@
+// +build unit integration
```
Comment (urso) [inline libbeat/outputs/elasticsearch/api_test.go]: we should not mix unit + integration tests in one file. Rather split into api_test.go and api_integration_test.go

**Comment by urso** [comment]
I'm in favor for using tags for integration tests, but I'm not happy about using tags for unit tests (as I very often do 'go test' in subdirectories). I'd rather have unit tests being executed always.

**Comment by tsg** [comment]
Ah, yes, I'd also prefer `go test` to do the unit tests.

File: libbeat/outputs/elasticsearch/api_test.go (hunk 3 above)
Comment (ruflin) [inline libbeat/outputs/elasticsearch/api_test.go]: The problem is that unit and integration tests share some helper functions which are inside api_test. I Agree that we should put these functions into an independent file in the future (testing.go?), but right now it is the simplest solution.

**Comment by ruflin** [comment]
Ok, will try to find a solution for this.

File: libbeat/outputs/elasticsearch/api_test.go (hunk 3 above)
Comment (urso) [inline libbeat/outputs/elasticsearch/api_test.go]: as all '_test.go' files are normally compiled when doing 'go test' I like to put common test helpers into something like `common_test.go` or `helpers_test.go` or `support_test.go` . Maybe have `api_common/helper/support_test.go` if helpers are 'local'. Some common scheme would make sense. Don't like big _test.go with loads of helper code I have search for the tests in first place.

Comment (ruflin) [inline libbeat/outputs/elasticsearch/api_test.go]: @urso My idea was testing.go per package. I assume the outcome would be the same as with `helper_test.go`.

**Comment by ruflin** [comment]
Just tried removing the `-unit` flag. The problem with this is that then integration tests again runs also the unit tests. I would prefer if this is not the case.

File: libbeat/outputs/elasticsearch/api_test.go (hunk 3 above)
Comment (urso) [inline libbeat/outputs/elasticsearch/api_test.go]: is testing.go always compiled or only by 'go test' ?

**Comment by urso** [comment]
With unit tests supposed to be 'fast', I don't mind running them with integration tests.

File: libbeat/outputs/elasticsearch/api_test.go (hunk 3 above)
Comment (ruflin) [inline libbeat/outputs/elasticsearch/api_test.go]: Always I would assume, would that be an issue? There is one advantage to testing.go: This would also work across packages. I have this issue for example in metricbeat. Several sub packages need the same test helper, so putting it in support_test.go does not work, as only the test files in the same package are compiled. That is why the test helper ended up in the main file: https://github.com/elastic/beats/blob/master/metricbeat/module/redis/redis.go

**Comment by ruflin** [comment]
@urso I would like to have a separate coverage report. Like this it is always mixed. Also in case we add for example benchmark tests, we get mixed reports and unit tests are run with the benchmark tests.

File: libbeat/outputs/elasticsearch/api_test.go (hunk 3 above)
Comment (urso) [inline libbeat/outputs/elasticsearch/api_test.go]: one solution is to provide a test support package. e.g [net/http/httptest](https://golang.org/pkg/net/http/httptest/).

Comment (ruflin) [inline libbeat/outputs/elasticsearch/api_test.go]: Seems to be overkill just for 1-2 functions per module.

**Comment by ruflin** [comment]
@tsg @urso Any further thoughts on this? I agree that it would be nice if `go test` could still be used as it is but I haven't found so far a nice solution. Will you also be happy with `go test -tags=unit`?

**Comment by tsg** [comment]
I didn't test but I think it's possible to use `+build !integration` instead of `+build unit` and then they shouldn't be built when integration is set. That's what we need, right?

**Comment by ruflin** [comment]
Yes that would work, but it would mean we would have to add `!other-test-type` for each test type we add.

Lets recap:

- Currently to only run unit tests, currently `go test -short` has to be used
- With the current PR `go test -tags=unit` would have to be used

This PR adds separation between test types which makes it possible to add addtional test types in the future and completely separates the running of tests (currenty if running integration-tests, unit tests are also run). Our daily use cases currently do not need a separation of the reports.

Running unit tests with `go test` instead of `go test -short` or `go test -unit` would be nice but all discussed solutions so far would not bring the separation between tests this PR was intended for. 

Moving forward I see two options:
- Close PR and stay with status quo
- Merge PR and use `go test -tags=unit` to only run unit tests. Side note: `make unit` does that too

**Comment by urso** [comment]
toplevel we've got our own TestMain, right? Can we play some tricks here?

**Comment by urso** [comment]
running unit vs. integration tests kinda makes the problem 'not simple'... maybe we're getting to clever here.

**Comment by tsg** [comment]
Personally I never felt good about requiring `-short` for running the unit tests. I think that if `go test` just works things are easier for the occasional contributor, there is less friction when quickly trying something in the terminal and it makes the life easier for e.g. editors that run the unit tests automatically.

But this is obviously something minor, and I'd rather merge the PR in its current form than close it. It's just not clear to me what is the downside of using `!integration` in the unit test files? Benchmarks are executed with `go test -bench`, so we don't need an extra tag for that. Do we envision any other types of tests?

**Comment by urso** [comment]
I agree with @tsg. I think even 'go test -bench' will run all unit tests before running the benchmarks.

**Comment by andrewkroh** [comment]
I'm in favor of `go test` and `go tests -tags !integration`. Mainly because I think running the unit tests in my IDE will get harder to do and I'll probably forget to add the `unit` tag when I add new test files.

I think new users with Go experience expect to be able to run `go build` and `go test` with our project.

**Comment by ruflin** [comment]
@tsg @andrewkroh I updated the code based on the comments above. Is that now as you expected? If yes, please let me rebase and squash before merging. Currently also Jenkins shows a race condition but it should not be related to any of the changes in this PR.

File: filebeat/input/file_other_test.go [hunk 4]
```diff
@@ -1,4 +1,4 @@
-// +build !windows
+// +build !windows unit
```
Comment (tsg) [inline filebeat/input/file_other_test.go]: Is this "unit" forgotten?

**Comment by tsg** [comment]
I like it like this, thanks @ruflin!

File: filebeat/input/file_other_test.go (hunk 4 above)
Comment (ruflin) [inline filebeat/input/file_other_test.go]: @tsg yes, will fix it

**Comment by ruflin** [comment]
It seems that now when we run the tests a little bit differently, new race conditions are discovered. @andrewkroh Could you hlep me here? http://build-eu-00.elastic.co/job/Beats-PR/809/console

**Comment by andrewkroh** [comment]
I will take a closer look at the stack traces and see if I can identify the root-cause.

**Comment by andrewkroh** [comment]
I wasn't able to reproduce this on my machine or on the hetzner machine using your branch.

Unrelated, I did notice that when `make integration-tests-environment` runs that the `RACE_DETECTOR=1` environment variable isn't honored because the tests run inide the container (I think that's what happens).

**Comment by ruflin** [comment]
@andrewkroh I fixed the RACE_DECTOR for make integration-tests-environment.

If I run it locally, I still get some race conditions. I would still merge it as it is master and it will allow us to better track down and fix the race condition. WDYT?

**Comment by andrewkroh** [comment]
There's a failure in this [AppVeyor build](https://ci.appveyor.com/project/elastic-beats/beats/build/1690/job/c3u02hg7ix54c1ij).

If you hit the race conditions locally on OSX then maybe I did something wrong. Let me try again. I'd rather fix it before merging.
